Source/VMware.vSphereDSC/DSCResources/VMHostAgentVM.ps1

```diff
@@ -0,0 +1,266 @@
+        }
+    }
+
+    <#
+    .DESCRIPTION
+
+    Checks if the AgentVM Setting needs to be updated with the desired value.
+    #>
+    [bool] ShouldUpdateAgentVMSetting($esxAgentHostManager, $agentVmSetting, $agentVmSettingName, $getAgentVmSettingAsViewObjectMethodName) {
+        if ($null -eq $agentVmSetting) {
+            if ($null -ne $esxAgentHostManager.ConfigInfo.$agentVmSettingName) {
+                return $true
+            }
+        }
+        else {
+            if ($null -eq $esxAgentHostManager.ConfigInfo.$agentVmSettingName) {
+                return $true
+            }
+            else {
+                $agentVmSettingAsViewObject = $this.$getAgentVmSettingAsViewObjectMethodName($esxAgentHostManager)
+                if ($agentVmSetting -ne $agentVmSettingAsViewObject.Name) {
+                    return $true
+                }
+            }
+        }
+
+        return $false
+    }
+
+    <#
+    .DESCRIPTION
+
+    Checks if the AgentVM Configuration needs to be updated with the desired values.
+    #>
+    [bool] ShouldUpdateAgentVMConfiguration($esxAgentHostManager) {
+        if ($this.ShouldUpdateAgentVMSetting($esxAgentHostManager, $this.AgentVmDatastore, $this.AgentVmDatastoreName, $this.GetAgentVmDatastoreAsViewObjectMethodName)) {
+            return $true
+        }
+        elseif ($this.ShouldUpdateAgentVMSetting($esxAgentHostManager, $this.AgentVmNetwork, $this.AgentVmNetworkName, $this.GetAgentVmNetworkAsViewObjectMethodName)) {
+            return $true
+        }
+        else {
+            return $false
+        }
+    }
+
+    <#
+    .DESCRIPTION
+
+    Retrieves the Datastore for AgentVM from the server if it exists.
+    If the Datastore name is not passed it returns $null and if the Datastore does not exist
+    it throws an exception.
+    #>
+    [PSObject] GetDatastoreForAgentVM($vmHost) {
+        if ($null -eq $this.AgentVmDatastore) {
+            return $null
+        }
+
+        try {
+            $datastore = Get-Datastore -Server $this.Connection -Name $this.AgentVmDatastore -RelatedObject $vmHost -ErrorAction Stop
+            return $datastore.ExtensionData.MoRef
+        }
+        catch {
+            throw "Could not retrieve Datastore $($this.AgentVmDatastore) for VMHost $($vmHost.Name). For more information: $($_.Exception.Message)"
+        }
+    }
+
+    <#
+    .DESCRIPTION
+
+    Retrieves the Network for AgentVM from the specified VMHost if it exists.
+    If the Network name is not passed it returns $null and if the Network does not exist
+    it throws an exception.
+    #>
+    [PSObject] GetNetworkForAgentVM($vmHost) {
+        if ($null -eq $this.AgentVmNetwork) {
+            return $null
+        }
+
+        $foundNetwork = $null
+        $networks = $vmHost.ExtensionData.Network
+
+        foreach ($network in $networks) {
+            $networkAsViewObject = Get-View -Server $this.Connection -Id $network
```

Comment:
When Connect is VC you can't be sure the network returned by name is available for the ESX under configuration. You have to retrieve the network similiar to how you retrieve the datastore. For datastore you're using -RelatedObject. I don't think this is availablr for Get-VirtualNetwork, but you can use Get-VirtualProgroup for standard networks and Get-VDPortGroup by host for distributed networks. I'm not sure whether NSX networks are supported for agent VM, this would be good to check and figure out how to support them as well. They are exposed as OpaqueNetworks in VC.

It would be good to add such integrations tests with two hosts, datastore and network that are not available for one of them and try to update AgentVM Configuration for the one that has no access to the datastore and network.

Reply:
$vmHost.ExtensionData.Network is a collection of references to the subset of network objects in the datacenter that are available in this HostSystem. So the returned network should be available for the ESXi.